File: 0039-combination-sum/solution.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        def bfs(index, path, remaining):
            if remaining == 0:
                res.append(path[:])
            for cur in range(index, len(candidates)):
                if candidates[cur] > remaining:
                    continue
                path.append(candidates[cur])
                bfs(cur, path, remaining - candidates[cur])
                path.pop()
        bfs(0, [], target)
        return res

File: 0039-combination-sum/solution.cpp
class Solution {
public:
    void backtrack(int index, int remaining, vector<int>& path, const vector<int>& nums, vector<vector<int>>& res){
        if(remaining == 0){
            res.push_back(path);
            return;
        }
        if(remaining < 0){
            return;
        }
        for(int i = index; i < nums.size(); ++i){

            path.push_back(nums[i]);
            backtrack(i, remaining - nums[i], path, nums, res);
            path.pop_back();
            //backtrack(index, remaining - nums[i], path, nums, res);
        }

    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        //sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> path;

        backtrack(0, target, path, candidates, res);
        return res;
    }
};
